src/nn/tensor_collection/collection.rs

```diff
@@ -1,28 +1,62 @@
 #![allow(clippy::type_complexity)]
 
 use crate::{
+    prelude::Device,
     shapes::{Dtype, Shape},
-    tensor::{DeviceStorage, OneFillStorage, Tensor, ZeroFillStorage},
+    tensor::{OneFillStorage, Tensor, ZeroFillStorage},
 };
 
+use super::visitor::TensorFunction;
+
+#[macro_export]
+macro_rules! try_option {
+    ($e:expr) => {
+        if let Some(x) = $e {
+            x
+        } else {
+            return Ok(None);
+        }
+    };
+}
+
+pub type ModuleVisitorOutput<F, M, E, D> = Result<
+    Option<
+        <M as TensorCollection<E, D>>::Output<
+            <F as TensorFunction<E, D>>::OutE,
+            <F as TensorFunction<E, D>>::OutD,
+        >,
+    >,
+    <F as TensorFunction<E, D>>::Err,
+>;
+
+pub type TensorFunctionOutput<F, S, E, D> = Result<
+    Option<Tensor<S, <F as TensorFunction<E, D>>::OutE, <F as TensorFunction<E, D>>::OutD>>,
+    <F as TensorFunction<E, D>>::Err,
+>;
+
 /// A collection of named tensors. Implementing this trait will enable anything
 /// that operates on tensors, like resetting, EMA, counting number of params,
 /// gradient updates, etc.
-pub trait TensorCollection<E: Dtype, D: DeviceStorage>: Sized {
-    fn iter_tensors<V: ModuleVisitor<Self, E, D>>(visitor: &mut V) -> Result<(), V::Err>;
+pub trait TensorCollection<E: Dtype, D: Device<E>>: Sized {
+    type Output<E2: Dtype, D2: Device<E2>>;
+
+    fn iter_tensors<V: ModuleVisitor<Self, E, D>>(
+        visitor: &mut V,
+    ) -> ModuleVisitorOutput<V::Func, Self, E, D>;
```

Comment:
I guess to support ToDtype & ToDevice we'd need different generics for dtype & device here?
Something like?
```suggestion
    fn iter_tensors<E2, D2, V: ModuleVisitor<Self, E2, D2>>(
        visitor: &mut V,
    ) -> ModuleVisitorOutput<V::Func, Self, E2, D2>;
```

Reply:
Yes, I was using associated types of TensorFuncitons to specify the output dtype and device, but I have moved over to using type parameters.